test/hyperbahn-client/discover.js

```diff
@@ -0,0 +1,133 @@
+
+if (require.main === module) {
+    runTests(require('../lib/test-cluster.js'));
+}
+
+function runTests(HyperbahnCluster) {
+    HyperbahnCluster.test('discover success', {
+        size: 5
+    }, function t(cluster, assert) {
+        var bob = cluster.remotes.bob;
+        var steve = cluster.remotes.steve;
+
+        var discoverClient;
+
+        var client = new HyperbahnClient({
+            serviceName: 'hello-bob',
+            callerName: 'hello-bob-test',
+            hostPortList: cluster.hostPortList,
+            tchannel: bob.channel,
+            logger: DebugLogtron('hyperbahnClient')
+        });
+
+        client.once('advertised', onAdvertised);
+        client.advertise();
+
+        function onAdvertised() {
+            discoverClient = new HyperbahnClient({
+                serviceName: 'hello-steve',
+                callerName: 'hello-steve-test',
+                hostPortList: cluster.hostPortList,
+                tchannel: steve.channel,
+                logger: DebugLogtron('hyperbahnClient')
+            });
+
+            discoverClient.discover({serviceName: 'hello-bob'}, onDiscovered);
+        }
+
+        function onDiscovered(err, hosts) {
+            assert.error(err, "foo", err);  // Should not have error here
+
+            assert.deepLooseEqual(hosts, [bob.channel.hostPort]);
+
+            client.destroy();
+            discoverClient.destroy();
+            assert.end();
+        }
+    });
+
+    HyperbahnCluster.test('discover no peers', {
+        size: 5
+    }, function t(cluster, assert) {
+        var steve = cluster.remotes.steve;
+
+        var discoverClient = new HyperbahnClient({
+            serviceName: 'hello-steve',
+            callerName: 'hello-steve-test',
+            hostPortList: cluster.hostPortList,
+            tchannel: steve.channel,
+            logger: DebugLogtron('hyperbahnClient')
+        });
+
+        discoverClient.discover(null, onDiscovered);
+
+        function onDiscovered(err, hosts) {
+            assert.error(!err, "NoPeersAvailable response", "expected NoPeersAvailable error");
+            assert.deepEqual(hosts, [], "expect empty host list");
+
+            discoverClient.destroy();
+            assert.end();
+        }
+    });
+
+    HyperbahnCluster.test('discover hard tchannel error', {
+        size: 5
+    }, function t(cluster, assert) {
+        var steve = cluster.dummies[1];
+
+        var discoverClient = new HyperbahnClient({
+            serviceName: 'hello-steve',
+            callerName: 'hello-steve-test',
+            hostPortList: [steve.hostPort],
+            tchannel: steve,
+            // Use a null logger here because the debugger one throws an
+            // exception when an error is logged.
+            logger: NullLogtron('hyperbahnClient')
+        });
+
+        discoverClient.discover({timeout: 100}, onDiscovered);
+
+        function onDiscovered(err, hosts) {
+            assert.error(!err, "expect TChannel error");
```

Comment:
Same feedback as above, having two assertion lines helps with reading the tests.